devito/passes/iet/langbase.py

```diff
@@ -359,3 +316,52 @@ def _is_offloadable(self, iet):
         buffers = [f for f in functions if f.is_Array and f._mem_mapped]
         hostfuncs = [f for f in functions if not is_on_device(f, self.gpu_fit)]
         return not (buffers and hostfuncs)
+
+
+@singledispatch
+def infer_transfer_datashape(f, *args):
+    """
+    Return the best shape to efficiently transfer `f` between the host
+    and a device.
+
+    First of all, we observe that the minimum shape is not necessarily the
+    best shape, because data contiguity plays a role.
+
+    So even in the most simple case, we transfer *both* the DOMAIN and the
+    HALO region. Consider the following:
+
+      .. code-block:: C
+
+        for (int x = x_m; x <= x_M; x += 1)
+          for (int y = y_m; y <= y_M; y += 1)
+            usaveb0[0][x + 1][y + 1] = u[t1][x + 1][y + 1];
+
+    Here the minimum shape to transfer `usaveb0` would be
+    `(1, x_M - x_m + 1, y_M - y_m + 1) == (1, x_size, y_size)`, but
+    we would rather transfer the contiguous chunk that also includes the
+    HALO region, namely `(1, x_size + 2, y_size + 2)`.
+
+    Likewise, in the case of SubDomains/SubDimensions:
+
+      .. code-block:: C
+
+        for (int xi = x_m + xi_ltkn; xi <= x_M - xi_rtkn; xi += 1)
+          for (int yi = y_m + yi_ltkn; yi <= y_M - yi_rtkn; yi += 1)
+            usaveb0[0][xi + 1][yi + 1] = u[t1][xi + 1][yi + 1];
+
+    We will transfer `(1, x_size + 2, y_size + 2)`.
+
+    In the future, this behaviour may change, or be made more sophisticated.
```

Comment:
Grammarly (some commas..etc..):
```
    In the future, this behaviour may change or be made more sophisticated.
    Note that any departure from this simple heuristic will require non-trivial
    additions to the compilation toolchain. For example, take the SubDomain
    example above. If we wanted to transfer the minimum shape, that is
    `(1, x_M - x_m - xi_ltkn - xi_rtkn + 1, y_M - y_m - yi_ltkn - yi_rtkn + 1)`,
    we would need to track both the iteration space of the computation and
    the write-to offsets (e.g., `xi + 1` and `yi + 1` in `usaveb0[0][xi + 1][yi + 1]`)
    because clearly, we would need to transfer the right amount starting at the
    right offset.
```